cognite/powerops/resync/validation.py

```diff
@@ -0,0 +1,209 @@
+    for model in models:
+        base_mappings = find_mappings(model, "base")
+        for mapping in base_mappings:
+            for validation_range_str, validation_range in validation_ranges.items():
+                ts_validation = TimeSeriesValidation(
+                    mapping=mapping,
+                    validation_range=validation_range,
+                )
+                if model.model_name not in ts_validation.data_models:
+                    ts_validation.data_models.append(model.model_name)
+                ts_validations[validation_range_str][mapping.external_id] = ts_validation
+
+    # Get any additional timeseries that are part of the models but are not part of the mappings.
+    # These are ambiguous... TODO Overwrite previous mappings or not? These are hard to connect back to SHOP path.
+    additional_timeseries = [ts for model in models for ts in model.timeseries()]
+    for additional_ts in additional_timeseries:
+        for validation_range_str in validation_ranges:
+            ts_validations[validation_range_str][require(additional_ts.external_id)] = TimeSeriesValidation(
+                validation_range=validation_ranges[validation_range_str],
+                mapping=InputTimeSeriesMappingApply(  # dummy mapping, for validation only
+                    external_id=require(additional_ts.external_id),
+                    cdf_time_series=require(additional_ts.external_id),
+                    retrieve="RANGE",
+                    shop_attribute_name="N/A",  # could get this from the model, possibly enhance .timeseries()
+                ),
+            )
+
+    return ts_validations, validation_ranges
+
+
+def perform_validation(
+    po_client: PowerOpsClient, ts_validations: PreparedValidationsT, validation_ranges: ValidationRangesT
+) -> None:
+    for range_str, validations_in_range in ts_validations.items():
+        validation_range = validation_ranges[range_str]
+        all_mappings = [validation.mapping for validation in validations_in_range.values()]
+        ts_mappings = list(filter(lambda row: row.cdf_time_series, all_mappings))
+
+        logger.info(f"Retrieving datapoints for {len(ts_mappings)} mappings from range {range_str}")
+        datapoints = retrieve_time_series_datapoints(
+            po_client.cdf,
+            ts_mappings,
+            start=arrow_to_ms(validation_range.start),
+            end=arrow_to_ms(validation_range.end),
+        )
+        warnings_n = 0
+        for ts_mapping in ts_mappings:
+            data: pd.Series = datapoints.get(require(ts_mapping.cdf_time_series), pd.Series())
+            if not len(data):
+                ts_validation: TimeSeriesValidation = ts_validations[range_str][ts_mapping.external_id]
+                logger.warning(
+                    f"No datapoints found for range {range_str} in timeseries '{ts_mapping.cdf_time_series}',"
+                    f" used in models: {', '.join(ts_validation.data_models)}."
+                )
```

Comment:
Up for discussion, but in `recyns` I use a `echo` callable with `echo("some message")` and echo("some warning", is_warning=True)` to be able to use `typer` native way of printing to terminal. Suggest passing that in here as well and replace the logger calls.